kstreams/middleware/middleware.py

```diff
@@ -0,0 +1,72 @@
+import logging
+import typing
+
+from aiokafka import errors
+
+from kstreams import ConsumerRecord, types
+
+if typing.TYPE_CHECKING:
+    from kstreams import Stream  #  pragma: no cover
+
+
+logger = logging.getLogger(__name__)
+
+
+class Middleware(typing.Protocol):
+    def __init__(
+        self,
+        *,
+        next_call: types.NextMiddlewareCall,
+        send: types.Send,
+        stream: "Stream",
+        **kwargs: typing.Any,
+    ) -> None:
+        ...  #  pragma: no cover
+
+    async def __call__(self, cr: ConsumerRecord) -> typing.Any:
+        ...  #  pragma: no cover
+
+
+class MiddlewareFactory:
```

Comment:
nitpick: Should we call this `Middleware` instead? 

Factory is confusing to me, it is a factory, but for me, in a factory pattern, I expect the factory class to internally have access to all the classes.

Something like this:
```
class JsonSerializer:
    def __init__(self):
        pass

    def serialize(self, data):
        return data

class AvroSerializer:
    def __init__(self):
        pass

    def serialize(self, data):
        return data
    
class SerializerFactory:
    def __init__(self, format: str):
        if format == "json":
            self.serializer = JsonSerializer()
        elif format == "avro":
            self.serializer = AvroSerializer()
    
    def serialize(self, data):
        return self.serializer.serialize(data)
```